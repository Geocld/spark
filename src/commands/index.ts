export * from './info'
export * from './init-cliff'
export * from './init'
export * from './log'
export * from './publish'
export * from './run'
