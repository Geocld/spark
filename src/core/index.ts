export * from './info'
export * from './init'
export * from './init-cliff'
export * from './log'
export * from './publish'
export * from './run'
