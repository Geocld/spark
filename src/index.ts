#!/usr/bin/env node

import yargs from 'yargs/yargs'
import { hideBin } from 'yargs/helpers'
import init from './core/init'
import run from './core/run'
import publish from './core/publish'
import info from './core/info'

type Option = boolean | undefined

yargs(hideBin(process.argv))
  .usage('Usage: $0 <command> [options]')
  .command(
    'init',
    'Create a new monorepo or upgrade an existing repo to monorepo.',
    () => {},
    (argv) => {
      init()
    }
  )
  .command(
    'info',
    'Prints information about the monorepo.',
    (args) => {
      args.option('tree', { demand: false }).alias('t', 'tree')
    },
    (argv) => {
      const tree = argv.tree as Option
      info(tree)
    }
  )
  .command(
    'run',
    'Run the script of package.',
    () => {},
    (argv) => {
      run()
    }
  )
  .command(
    'publish',
    'Publish packages in the current project.',
    (args) => {
      args.option('force', { demand: false }).alias('f', 'force')
      args.option('noPublish', { demand: false }).alias('np', 'noPublish')
    },
    (argv) => {
      const force = argv.force as Option
      const noPublish = argv.noPublish as Option
      publish(force, noPublish)
    }
  )
  .demandCommand(1, 'A command is required. Pass --help to see all available commands and options.')
  .strict()
  .alias('h', 'help')
  .alias('v', 'version')
  .parse()
