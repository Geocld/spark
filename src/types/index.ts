export interface SparkeeConfig {
  packages: string[]
  singleRepo: boolean
  moduleManager: 'npm' | 'yarn' | 'pnpm'
  logPresetTypes: LogPreset[]
  logCommit: {
    type: string
    commitTag: string
  }
}

// types: https://github.com/conventional-changelog/conventional-changelog-config-spec/blob/master/versions/2.2.0/README.md#type
export type LogPreset = {
  type: string
  scope?: string
  section?: string
  hidden?: boolean
}

export interface PnpmWorkspace {
  packages: string[]
}

export interface PackageJson {
  name: string
  version?: string
  scripts?: ScriptsMap
  dependencies?: DependencyMap
}

export interface WorkspacePackage extends PackageJson {
  path: string
  pkg?: PackageJson
}

export type WorkspacePackages = WorkspacePackage[]

export type WorkspacePackageWithoutPkg = Omit<WorkspacePackage, 'pkg'>

export interface ScriptsMap {
  [scriptName: string]: string
}

export interface DependencyMap {
  [depName: string]: string
}
