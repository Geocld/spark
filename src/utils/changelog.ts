import fs from 'fs'
import { join } from 'path'
import shell from 'shelljs'
import consola from 'consola'
import chalk from 'chalk'
import dayjs from 'dayjs'
import { LOCAL_CLIFF_TOML, DEFAULT_MONOREPO_CLIFF_TOML, DEFAULT_SINGLEREPO_CLIFF_TOML } from '../common/constans'
import type { WorkspacePackageWithoutPkg } from './../types'

function execCommand(args: Array<string|undefined>, message?: string) {
  const { code, stderr } = shell.exec(args.join(' '), { silent: true })
  if (code !== 0) {
    consola.error(stderr)
  } else {
    message && consola.success(chalk.green(message))
  }
}

export function generateChangeLog(pkg: WorkspacePackageWithoutPkg, singleRepo = false) {
  // Get local sparkee-cliff.toml otherwise use config inside.
  const { name, path, version } = pkg
  let cliffToml = LOCAL_CLIFF_TOML
  if (!fs.existsSync(LOCAL_CLIFF_TOML)) {
    consola.warn('Local sparkee-cliff.toml not found, use default Config.')
    cliffToml = singleRepo ? DEFAULT_SINGLEREPO_CLIFF_TOML : DEFAULT_MONOREPO_CLIFF_TOML
  }

  const changelogFile = singleRepo ? './CHANGELOG.md' : join(path, './CHANGELOG.md')
  
  let execChangeLogCommand = [
    'npx', 
    'git-cliff', 
    '--config', 
    cliffToml, 
    '-o', 
    changelogFile
  ]

  if (!singleRepo) {
    // TODO: custom template in monorepo
    execChangeLogCommand = execChangeLogCommand.concat([
      '--include-path',
      `"${path}/**"`,
      '--body', // Sets the template for the changelog body
      `"""{% if version %}
        {% if version is matching(\\"^${name}.*\\") %}
          ## [{{ version }}] - {{ timestamp | date(format=\\"%Y-%m-%d\\") }}
        {% endif %}
      {% else %}\
## [${name}@${version}] - ${dayjs().format('YYYY-MM-DD')} \n
      {% endif %}\
      {% for group, commits in commits | group_by(attribute=\\"group\\") %}
          ### {{ group | upper_first }}
          {% for commit in commits %}
              - {% if commit.breaking %}[**breaking**] {% endif %}{{ commit.message | upper_first }} {% if commit.id %}([{{ commit.id | truncate(length=7, end=\\"\\") }}]({{ commit.id }})){% endif %}\
          {% endfor %}
      {% endfor %}\n"""`
    ])
  }

  execCommand(execChangeLogCommand, 'Generate CHANGELOG.md successful.')

  // Replace [unreleased] in singleRepo CHANGELOG.md
  if (singleRepo) {
    let logContent = fs.readFileSync(changelogFile, 'utf-8')
    logContent = logContent.replace('[unreleased]', `[${version}] - ${dayjs().format('YYYY-MM-DD')}`)
    fs.writeFileSync(changelogFile, logContent)
  }
}