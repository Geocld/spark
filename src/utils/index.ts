import { PNPM_WORKSPACE, ROOT, ROOT_PACKAGE, SPARK_JSON } from '../common/constans'
import type { PackageJson, PnpmWorkspace, SparkeeConfig, WorkspacePackages } from '../types'
import chalk from 'chalk'
import consola from 'consola'
import globStandard from 'glob'
import jsonfile from 'jsonfile'
import { stat, writeFile } from 'node:fs/promises'
import { join } from 'path'
import readYamlFile from 'read-yaml-file'
import shell from 'shelljs'
import { promisify } from 'util'

const glob = promisify(globStandard)

// https://github.com/microsoft/TypeScript/issues/16069#issuecomment-369374214
export function isNotNullOrUndefined<T extends Object>(input: null | undefined | T): input is T {
  return input != null
}

export function getFirstRegexGroup(regexp: RegExp, url: string) {
  return Array.from(url.matchAll(regexp), (m) => m[1])
}

export const fileExists = async (path: string) => !!(await stat(path).catch((_e: unknown) => false))

// Get workspace folder, default is 'packages'
async function getWorkspaceFolders(packages: string[] | string = '*'): Promise<string[]> {
  let folders: string[] = []

  try {
    const pnpmWorkspace = await readYamlFile<PnpmWorkspace>(PNPM_WORKSPACE)
    let wPackages = pnpmWorkspace.packages

    wPackages = wPackages.map((wp) => {
      return `${wp.split('/')[0]}/*`
    })

    await Promise.all(
      wPackages.map(async (wp) => {
        const wFolders = await glob(wp)
        folders = folders.concat(wFolders)
      })
    )
  } catch {
    folders = await glob('packages/*')
  }

  if (packages === '*') {
    return folders
  }

  // All packages match the packages of (spark.json -> packages)
  const _folders = await Promise.all(
    folders.map(async (folder) => {
      if (!(await stat(folder)).isDirectory()) return null

      const pkg = await getPackageJson(folder)
      if (packages.indexOf(pkg.name) === -1) return null
      return folder
    })
  )

  return _folders.filter(isNotNullOrUndefined)
}

export function exit() {
  process.exit(1)
}

export function formatStdout(stdout: string): string {
  return stdout.trim().replace('\n', '')
}

// Get all workspace packages
export async function getWorkspacePackages(): Promise<PackageJson[]> {
  const folders: string[] = await getWorkspaceFolders()

  // If no workspace folder is found, cause of missing configuration, return root package.json
  if (!folders.length) {
    const pkg = await getPackageJson()
    return [pkg]
  }

  const pkgs = await Promise.all(
    folders.map(async (folder) => {
      if (!(await stat(folder)).isDirectory()) return null

      const pkg = await getPackageJson(folder)
      return pkg
    })
  )

  return pkgs.filter(isNotNullOrUndefined)
}

// Get a property from all workspace packages
export async function getPkgsProperty(property: string): Promise<Array<keyof PackageJson>> {
  const pkgs = await getWorkspacePackages()
  const properties: Array<keyof PackageJson> = pkgs.map((pkg) => {
    if (!pkg[property]) return null
    return pkg[property]
  })

  return properties.filter((p) => p)
}

export async function exec(
  cmd: string,
  silent: boolean = true
): Promise<{ stdout: string; stderr: string; code: number }> {
  return shell.exec(cmd, { silent })
}

export function step(msg: string) {
  consola.log(chalk.cyan(msg))
}

export async function getChangedPackages(force: boolean = false): Promise<WorkspacePackages> {
  let lastTag: string

  // get packages of spark.json
  if (!(await fileExists(SPARK_JSON))) {
    consola.error('`spark.json` was not found, please try to run `sparkee init` to init repo.')
    exit()
  }
  const sparkConfig = await getSparkeeConfig()
  const { singleRepo, packages } = sparkConfig

  const { stdout: tag, stderr } = await exec('git describe --tags --abbrev=0')

  if (stderr) {
    !force && consola.warn(`Couldn't get the last tag, using first commit...`)
    const { stdout, stderr } = await exec('git rev-list --max-parents=0 HEAD')
    lastTag = formatStdout(stdout)
  } else {
    lastTag = formatStdout(tag)
  }

  if (singleRepo) {
    consola.warn('You will publish as singleRepo.')
    const { stdout: hasChanges } = await exec(`git diff ${lastTag}`)
    const pkg = await getPackageJson()
    if (!(force || hasChanges)) {
      return []
    }
    return [
      {
        path: ROOT,
        name: pkg.name,
        version: pkg.version,
        pkg,
      },
    ]
  } else {
    // monorepo
    // get folders match managed packages
    const folders: string[] = await getWorkspaceFolders(packages)

    const pkgs = await Promise.all(
      folders.map(async (folder) => {
        const pkg = await getPackageJson(folder)
        const { stdout: hasChanges } = await exec(
          `git diff ${lastTag} -- ${join(folder, 'src')} ${join(folder, 'package.json')}`
        )
        if (force || hasChanges) {
          // force mode return all packages
          return {
            path: folder,
            name: pkg.name,
            version: pkg.version,
            pkg,
          }
        }
      })
    )

    return pkgs.filter(isNotNullOrUndefined)
  }
}

export async function updateVersions(packageList: WorkspacePackages): Promise<void[]> {
  return Promise.all(
    packageList.map(async ({ pkg, version, path }) => {
      if (!pkg) {
        consola.error('Packages can not be empty!')
        return exit()
      }

      pkg.version = version
      const content = JSON.stringify(pkg, null, 2) + '\n'

      return await writeFile(join(path, 'package.json'), content, { encoding: 'utf8' })
    })
  )
}

export async function runTaskSync(tasks: Function[]): Promise<any[]> {
  for (const task of tasks) {
    if (typeof task !== 'function') {
      throw new TypeError(`Expected task to be a \`Function\`, received \`${typeof task}\``)
    }
  }

  const results: Function[] = []

  for (const task of tasks) {
    results.push(await task()) // eslint-disable-line no-await-in-loop
  }

  return results
}

export async function getSparkeeConfig(): Promise<SparkeeConfig> {
  const config = (await jsonfile.readFile(SPARK_JSON)) as SparkeeConfig
  return config
}

export async function getPackageJson(packageJsonPath?: string): Promise<PackageJson> {
  const pkg = (await jsonfile.readFile(
    packageJsonPath ? join(packageJsonPath, 'package.json') : ROOT_PACKAGE
  )) as PackageJson
  return pkg
}
