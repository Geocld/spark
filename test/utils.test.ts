import { getPackageJson, getPkgsProperty, getSparkeeConfig, getWorkspacePackages } from '../src/utils'
import { describe, expect, it } from 'vitest'

import packageJson from '../package.json'

describe('utils test', async () => {
  it('should load spark.json config file', async () => {
    const sparkeeConfig = await getSparkeeConfig()

    expect(sparkeeConfig).toEqual({
      $schema: './template/schema.json',
      singleRepo: true,
      moduleManager: 'npm',
    })
  })

  it('should load root package.json file', async () => {
    const rootPackageJson = await getPackageJson()

    expect(rootPackageJson.name).toEqual(packageJson.name)
    expect(rootPackageJson.license).toEqual(packageJson.license)
  })

  it('should load singlerepo root package.json file', async () => {
    const allPackages = await getWorkspacePackages()

    expect(allPackages).toEqual([packageJson])
  })

  it('should load root package.json scripts', async () => {
    const allPackages = await getPkgsProperty('scripts')

    expect(allPackages).toEqual([packageJson.scripts])
  })
})
