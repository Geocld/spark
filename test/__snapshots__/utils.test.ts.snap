// Vitest Snapshot v1

exports[`utils test > should load root package.json scripts 1`] = `
[
  {
    "build": "rimraf dist && tsc",
    "coverage": "vitest run --coverage",
    "dev": "rimraf dist && tsc -w",
    "fmt": "rome format src --write",
    "lint": "rome check src",
    "pub": "node dist/index.js publish",
    "test": "vitest",
  },
]
`;

exports[`utils test > should load singlerepo root package.json file 1`] = `
[
  {
    "author": "lijiahao <lijiahao5372@gmail.com>",
    "bin": {
      "sparkee": "dist/index.js",
      "spk": "dist/index.js",
    },
    "dependencies": {
      "chalk": "^4.1.2",
      "consola": "^2.15.3",
      "dayjs": "^1.11.7",
      "fs-extra": "^10.1.0",
      "git-cliff": "1.1.2",
      "glob": "^8.0.3",
      "inquirer": "^8.2.4",
      "jsonfile": "^6.1.0",
      "read-yaml-file": "^2.1.0",
      "rimraf": "^3.0.2",
      "semver": "^7.3.7",
      "shelljs": "^0.8.5",
      "shelljs-live": "^0.0.5",
      "yargs": "^17.5.1",
    },
    "description": "A publish tool for monorepo,just like lerna but support pnpm.",
    "devDependencies": {
      "@types/fs-extra": "^11.0.1",
      "@types/glob": "^8.0.1",
      "@types/jsonfile": "^6.1.1",
      "@types/node": "^17.0.43",
      "@types/shelljs": "^0.8.11",
      "@types/yargs": "^17.0.10",
      "rome": "^11.0.0",
      "typescript": "^4.9.4",
      "vitest": "^0.28.4",
    },
    "keywords": [
      "monorepo",
      "pnpm",
    ],
    "license": "MIT",
    "main": "dist/index.js",
    "name": "sparkee",
    "optionalDependencies": {
      "git-cliff-darwin-arm64": "1.1.2",
      "git-cliff-darwin-x64": "1.1.2",
      "git-cliff-linux-arm64": "1.1.2",
      "git-cliff-linux-x64": "1.1.2",
      "git-cliff-windows-arm64": "1.1.2",
      "git-cliff-windows-x64": "1.1.2",
    },
    "repository": {
      "type": "git",
      "url": "git+https://github.com/Geocld/sparkee",
    },
    "scripts": {
      "build": "rimraf dist && tsc",
      "coverage": "vitest run --coverage",
      "dev": "rimraf dist && tsc -w",
      "fmt": "rome format src --write",
      "lint": "rome check src",
      "pub": "node dist/index.js publish",
      "test": "vitest",
    },
    "version": "1.2.0",
  },
]
`;
